src/coreclr/jit/gentree.cpp

```diff
@@ -8600,6 +8600,46 @@ void Compiler::gtUpdateNodeOperSideEffects(GenTree* tree)
     }
 }
 
+//------------------------------------------------------------------------
+// gtUpdateNodeOperSideEffectsPost: Update the side effects based on the node operation,
+// in the post-order visit of a tree walk. It is expected that the pre-order visit cleared
+// the bits, so the post-order visit only sets them. This is important for binary nodes
+// where one child already may have set the GTF_EXCEPT bit. Note that `SetIndirExceptionFlags`
+// looks at its child, which is why we need to do this in a bottom-up walk.
+//
+// Arguments:
+//    tree            - Tree to update the side effects on
+//
+// Notes:
+//    This method currently only updates GTF_EXCEPT, GTF_ASG, and GTF_CALL flags.
+//    The other side effect flags may remain unnecessarily (conservatively) set.
+//    The caller of this method is expected to update the flags based on the children's flags.
+//
+void Compiler::gtUpdateNodeOperSideEffectsPost(GenTree* tree)
+{
+    if (tree->OperMayThrow(this))
+    {
+        tree->gtFlags |= GTF_EXCEPT;
+    }
+    else
+    {
+        if (tree->OperIsIndirOrArrLength())
+        {
+            tree->SetIndirExceptionFlags(this);
```

Comment:
It looks like we do the same that we do in `if (tree->OperMayThrow(this))`:
that is what happens in `if (tree->OperMayThrow(this))`
https://github.com/dotnet/runtime/blob/5b0c6dd5756eea63a57b1f69b182f7386da0f631/src/coreclr/jit/gentree.cpp#L5792-L5794

and that is what happens in `SetIndirExceptionFlags`:
https://github.com/dotnet/runtime/blob/5b0c6dd5756eea63a57b1f69b182f7386da0f631/src/coreclr/jit/gentree.cpp#L9919-L9928

and I can't figure out what is the difference and what is responsible for what.

Reply:
yeah, it's quite confusing what the contract is here. I can't clear it up, honestly. It seems that OperMayThrow is trying to figure out whether to set `GTF_EXCEPT` "from basic principles", that is, not depending on anyone having pre-set that bit. However, the `GT_INTRINSIC` case does consult that (maybe that's actually a bug). `fgAddrCouldBeNull` does walk through COMMAs. `SetIndirExceptionFlags` will *also* set GTF_EXCEPT if the addr has the `GTF_EXCEPT` bit set, essentially propagating it from it's child. However, it *doesn't* walk through COMMAs. And it's not clear who is expected to have set those bits.

I think the idea of `SetIndirExceptionFlags` is to call it when you create a new indir/arr_length node and all the children are already created, and have proper bits. The case in `gtUpdateNodeOperSideEffectsPost` (that I added) maybe isn't necessary at all: if we're walking up in post-order, `OperMayThrow` will set the correct bits based on the indir/arr_length and addr; `fgUpdateSideEffectsPost` will update the node from any COMMA child with op1 with GTF_EXCEPT set (IND(COMMA(...something with GTF_EXCEPT,...))).  However, that's not even quite correct, because `SetIndirExceptionFlags` will clear GTF_EXCEPT (and set GTF_IND_NONFAULTING) based on addr.